infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -303,26 +333,40 @@ public void read(CorfuPayloadMsg<ReadRequest> msg, ChannelHandlerContext ctx, IS
     private void getKnownAddressesInRange(CorfuPayloadMsg<KnownAddressRequest> msg,
                                           ChannelHandlerContext ctx, IServerRouter r) {
 
-        KnownAddressRequest request = msg.getPayload();
+        Lock lock = resetLock.readLock();
+        lock.lock();
+
         try {
-            Set<Long> knownAddresses = streamLog
-                    .getKnownAddressesInRange(request.getStartRange(), request.getEndRange());
-            r.sendResponse(ctx, msg,
-                    CorfuMsgType.KNOWN_ADDRESS_RESPONSE.payloadMsg(knownAddresses));
-        } catch (DataCorruptionException dc) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_CORRUPTION
-                    .payloadMsg(Address.NON_ADDRESS));
-            return;
-        } catch (Exception e) {
-            handleException(e, ctx, msg, r);
+            KnownAddressRequest request = msg.getPayload();
+            try {
+                Set<Long> knownAddresses = streamLog
+                        .getKnownAddressesInRange(request.getStartRange(), request.getEndRange());
+                r.sendResponse(ctx, msg,
+                        CorfuMsgType.KNOWN_ADDRESS_RESPONSE.payloadMsg(knownAddresses));
+            } catch (DataCorruptionException dc) {
+                r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_CORRUPTION
+                        .payloadMsg(Address.NON_ADDRESS));
+            } catch (Exception e) {
+                handleException(e, ctx, msg, r);
+            }
+        } finally {
+            lock.unlock();
         }
     }
 
     @ServerHandler(type = CorfuMsgType.COMPACT_REQUEST)
     private void handleCompactRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
         log.debug("handleCompactRequest: received a compact request {}", msg);
-        streamLog.compact();
-        r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
+
+        Lock lock = resetLock.writeLock();
+        lock.lock();
+
```

Comment:
epoch check after lock.

Reply:
As discussed twice the check will crash most of the unit tests, here is the build with the change https://travis-ci.org/github/CorfuDB/CorfuDB/builds/716015026

Reply:
also see my comment https://github.com/CorfuDB/CorfuDB/pull/2661#issuecomment-667404526